Native onchain development - Updated hello world lesson (#401)
solana-foundation/developer-content: content/courses/native-onchain-development/hello-world-program.md

## Summary

- Native Solana programs have a single **entry point** to process instructions.
- A program processes an instruction using the **program_id**, a list of

>   **accounts**, and **instruction_data** included with the instruction.

Reviewer: Love the attention to detail here. 👍

Author: Thank you! @mikemaccana

## Lesson

<Callout type="info">

The following guide assumes you are familiar with Solana program basics. If not,
check out
[Introduction to Onchain Programming](/content/courses/onchain-development/intro-to-onchain.md).
</Callout>

This lesson will introduce you to writing and deploying a Solana program using
the Rust programming language without any framework. This approach offers
greater control but requires you to handle much of the foundational work of

> creating an on-chain program yourself.

Reviewer: 'onchain' is one word. See https://github.com/solana-foundation/developer-content/blob/main/CONTRIBUTING.md#prose and fix everywhere.

To avoid the distractions of setting up a local development environment, we'll
be using a browser-based IDE called Solana Playground.

### Rust Basics

Before diving into building our "Hello, world!" program, let’s review some Rust
basics. For a deeper dive into Rust, check out the
[Rust Language Book](https://doc.rust-lang.org/book/ch00-00-introduction.html).

#### Module System

Rust organizes code using what is collectively referred to as the “module
system.” This includes:

> - **Modules**: Separate code into logical units to provide isolated namespaces

Reviewer: ```suggestion
- **Modules**: Separates code into logical units to provide isolated namespaces
```

Reviewer: Small accident, but the rest of your writing is excellent.

  for organization, scope, and privacy of paths.
- **Crates**: Either a library or an executable program. The source code for a
  crate is usually subdivided into multiple modules.
- **Packages**: A collection of crates along with a manifest file that specifies
  metadata and dependencies between packages.

Throughout this lesson, we’ll focus on using crates and modules.

#### Paths and Scope

> Crates in Rust contain modules that define functionality, which can be shared
> across multiple projects. If we want to access an item within a module, we need
> to know its "path," similar to navigating a filesystem.

Reviewer: ```suggestion
Crates contain modules that can be shared
across multiple projects. If we want to access an item within a module, we need
to know its "path," similar to navigating a filesystem.
```

Reviewer: Yours was better that the original but we can go further.

The reference is guaranteed to point to a valid value of a particular type. The
action of creating a reference in Rust is called “borrowing.”

In this example, when the `process_instruction` function is called, a user must
pass in values for the required arguments. The `process_instruction` function
then references the values passed in by the user, guaranteeing that each value
is the correct data type specified in the function.

Additionally, note the brackets `[]` around `&[AccountInfo]` and `&[u8]`. This
indicates that the `accounts` and `instruction_data` arguments expect "slices"
of types `AccountInfo` and `u8`, respectively. A “slice” is similar to an array
(a collection of objects of the same type), except the length is not known at
compile time. In other words, the `accounts` and `instruction_data` arguments
expect inputs of unknown length.

```rust
pub fn process_instruction(
    program_id: &Pubkey,
    accounts: &[AccountInfo],
    instruction_data: &[u8],
)
```

We can also have our functions return values by declaring the return type using
an arrow -> after the function.

In the example below, the `process_instruction` function will now return a value

> of type `ProgramResult`. We’ll go over this in the next section.

Reviewer: ```suggestion
of type `ProgramResult`. We'll go over this in the next section.
```

Reviewer: I'm not totally sure there's any benefit to smart quotes. The rest of the code uses regular quotes so I think we should stick to that, or change it everywhere. Feel free to change my mind!

Author: This was from old content only, I didn't notice. I have changed it to regular quotes across the lesson.

};
```

- `AccountInfo` - A struct within the `account_info` module that allows us to
  access account information.
- `entrypoint` -A macro that declares the entry point of the program.
- `ProgramResult` - A type within the `entrypoint` module that returns either a
  `Result` or `ProgramError`.
- `Pubkey` - A struct within the `pubkey` module that allows us to access
  addresses as public keys.

> - `msg` - A macro that allows us to print messages to the program log.

Reviewer: `msg` and `Pubkey` and `ProgramResult` are good. But `AccountInfo` and `entrypoint` are recursive. What is an entrypoint? What is account information?

> `entrypoint` - A macro that declares the function that receives incoming instructions and passes them on the right instruction handler.
 > `AccountInfo` - A struct that allows us to access account information like...

For example. You're a great writer I'm sure you can do better!

Author: Added more details as per suggestion to explain better.

entrypoint!(process_instruction);

pub fn process_instruction(
    _program_id: &Pubkey,
    _accounts: &[AccountInfo],
    _instruction_data: &[u8],
) -> ProgramResult {
    msg!("Hello, world!");

    Ok(())
}
```

### 4. Build and Deploy

Now, let's build and deploy our program using Solana Playground.

![Gif Solana Playground Build and Deploy](/public/assets/courses/unboxed/hello-world-build-deploy.gif)

### 5. Invoke Program

Finally, let's invoke our program from the client side. The main focus of this
lesson is building our Solana program, so we've provided
[the client code to invoke our “Hello, world!” program](https://github.com/Unboxed-Software/solana-hello-world-client)
for you to download.

This code includes a `sayHello` helper function that constructs and submits the
transaction. In the `index.ts` file, you'll find a variable named `programId`.
Update this with the program ID of the “Hello, world!” program you just deployed
using Solana Playground.

```typescript
let programId = new web3.PublicKey("<YOUR_PROGRAM_ID>");
```

You can find the program ID on Solana Playground as shown below.

> ![Gif Solana Playground Program ID](/public/assets/courses/unboxed/hello-world-program-id.gif)

Reviewer: You might want to make a new version of this image (or an mp4) if Solana Playground looks out of date. You decide.

Author: I have not updated the new version as there is not much difference in UI and UX.

Next, install the Node modules by running `npm i`.

Afterwards, execute `npm start`. This command will:

1. Generate a new keypair and create a `.env` file if it doesn't already exist.

> 2. Airdrop some SOL on the Devnet.

Reviewer: ```suggestion
2. Airdrop some SOL onto this account on devnet.
```

not 'the devnet'.

3. Invoke the “Hello, world!” program.
4. Output a transaction URL that you can view on Solana Explorer.

Copy the transaction URL from the console into your browser. Scroll down to the
Program Instruction Logs section to see “Hello, world!” displayed.

> ![Screenshot Solana Explorer Program Log](/public/assets/courses/unboxed/hello-world-program-log.png)

Reviewer: Likewise you may want to update this screenshot if you think it's necessary.

Author: The same has not been updated to the new version as there is not much difference in UI and UX except for the addition of 2 new icons for tutorials and programs.

Congratulations! You've successfully built and deployed a Solana program!

## Challenge

Now it’s your turn to build something independently. Since we're starting with
very simple programs, your task will closely resemble what we've just created.
The goal is to practice writing the code from scratch without referencing prior
examples, so try to avoid copying and pasting.

1. Write a new program that uses the `msg!` macro to print your own custom
   message to the program log.
2. Build and deploy your program just like we did in the lab.
3. Invoke your newly deployed program and use Solana Explorer to confirm that
   your message was printed in the program log.

As always, feel free to get creative with these challenges and go beyond the

> basic instructions if you want—most importantly, have fun!

Reviewer: ```suggestion
basic instructions if you want — most importantly, have fun!
```
